patch/llpcPatchCheckShaderCache.cpp

```diff
@@ -0,0 +1,161 @@
+ ***********************************************************************************************************************
+ */
+#define DEBUG_TYPE "llpc-patch-check-shader-cache"
+
+#include "llvm/Support/Debug.h"
+
+#include "llpcPatchCheckShaderCache.h"
+#include "llpcPipelineShaders.h"
+
+using namespace llvm;
+using namespace Llpc;
+
+// =====================================================================================================================
+// Initializes static members.
+char PatchCheckShaderCache::ID = 0;
+
+namespace Llpc
+{
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for checking shader cache
+PatchCheckShaderCache* CreatePatchCheckShaderCache()
+{
+    return new PatchCheckShaderCache();
+}
+
+} // Llpc
+
+namespace {
+
+// =====================================================================================================================
+// Hash each map key and value
+template <class MapType>
+static void UpdateHashForMap(MapType& m, MD5* pHasher)
+{
+    size_t mapCount;
+    pHasher->update(ArrayRef<uint8_t>(reinterpret_cast<const uint8_t*>(&mapCount), sizeof(mapCount)));
+    for (auto mapIt : m)
+    {
+        pHasher->update(ArrayRef<uint8_t>(reinterpret_cast<const uint8_t*>(&mapIt.first), sizeof(mapIt.first)));
+        pHasher->update(ArrayRef<uint8_t>(reinterpret_cast<const uint8_t*>(&mapIt.second), sizeof(mapIt.second)));
+    }
+}
+
+} // anonymous
+
+// =====================================================================================================================
+PatchCheckShaderCache::PatchCheckShaderCache()
+    :
+    Patch(ID)
+{
+    initializePipelineShadersPass(*PassRegistry::getPassRegistry());
+    initializePatchCheckShaderCachePass(*PassRegistry::getPassRegistry());
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+bool PatchCheckShaderCache::runOnModule(
+    Module& module)  // [in,out] LLVM module to be run on
+{
+    LLVM_DEBUG(dbgs() << "Run the pass Patch-Check-Shader-Cache\n");
+
+    if (m_callbackFunc == nullptr)
+    {
+        // No shader cache in use.
+        return false;
+    }
+
+    Patch::Init(&module);
+
+    MD5::MD5Result inOutUsageHashes[ShaderStageGfxCount] = {};
+    ArrayRef<uint8_t> inOutUsageHashValues[ShaderStageGfxCount] = {};
+    auto stageMask = m_pContext->GetShaderStageMask();
+
+    // Build input/output layout hash per shader stage
+    for (auto stage = ShaderStageVertex; stage < ShaderStageGfxCount; stage = static_cast<ShaderStage>(stage + 1))
+    {
+        if ((stageMask & ShaderStageToMask(stage)) == 0)
+        {
+            continue;
+        }
+
+        auto pResUsage = m_pContext->GetShaderResourceUsage(stage);
+        MD5 hasher;
```

Comment:
why do you use MD5 hash instead of Metro hash?

Reply:
MD5 already exists in LLVM. We do not want unnecessary dependencies in the middle-end to anything outside itself or LLVM.

Reply:
PAL used MD5 hash before we switched to metro hash.
and whole vulkan and PAL use metro hash to calculate the hash code.
LLPC pipeline hash is based on Metro hash too. so please keep use metro hash in current phase, thanks!.

Reply:
I removed the hashing completely from the middle-end to avoid introducing a dependency on metrohash.